tap/xdp/ipproto.c

```diff
@@ -0,0 +1,79 @@
+// +build ignore
+
+#include <linux/bpf.h>
+#include <linux/in.h>
+#include <linux/udp.h>
+#include <linux/if_ether.h>
+#include <linux/ip.h>
+#include <linux/ipv6.h>
+#include <bpf/bpf_helpers.h>
+#include <bpf/bpf_endian.h>
+
+
+#define MAX_SOCKS 64
+
+static volatile unsigned const char PROTO;
+static volatile unsigned const char PROTO = IPPROTO_ICMP;
+
+//Ensure map references are available.
+/*
+        These will be initiated from go and
+        referenced in the end BPF opcodes by file descriptor
+*/
+
+struct bpf_map_def SEC("maps") xsks_map = {
+	.type = BPF_MAP_TYPE_XSKMAP,
+	.key_size = sizeof(int),
+	.value_size = sizeof(int),
+	.max_entries = MAX_SOCKS,
+};
+
+struct bpf_map_def SEC("maps") qidconf_map = {
+	.type = BPF_MAP_TYPE_ARRAY,
+	.key_size = sizeof(int),
+	.value_size = sizeof(int),
+	.max_entries = MAX_SOCKS,
+};
+
+
+SEC("xdp_sock") int xdp_sock_prog(struct xdp_md *ctx)
+{
+
+	int *qidconf, index = ctx->rx_queue_index;
+
+	// A set entry here means that the correspnding queue_id
+	// has an active AF_XDP socket bound to it.
+	qidconf = bpf_map_lookup_elem(&qidconf_map, &index);
+	if (!qidconf)
+		return XDP_PASS;
+
+	// redirect packets to an xdp socket that match the given IPv4 or IPv6 protocol; pass all other packets to the kernel
+	void *data = (void*)(long)ctx->data;
+	void *data_end = (void*)(long)ctx->data_end;
+	struct ethhdr *eth = data;
+	__u16 h_proto = eth->h_proto;
+	if ((void*)eth + sizeof(*eth) <= data_end) {
+		if (bpf_htons(h_proto) == ETH_P_IP) {
+			struct iphdr *ip = data + sizeof(*eth);
+			if ((void*)ip + sizeof(*ip) <= data_end) {
+				if (ip->protocol == PROTO) {
+					if (*qidconf)
+						return bpf_redirect_map(&xsks_map, index, 0);
+				}
+			}
+		} else if (bpf_htons(h_proto) == ETH_P_IPV6) {
+			struct ipv6hdr *ip = data + sizeof(*eth);
+			if ((void*)ip + sizeof(*ip) <= data_end) {
+				if (ip->nexthdr == PROTO) {
+					if (*qidconf)
+						return bpf_redirect_map(&xsks_map, index, 0);
```

Comment:
Won't this "hijack" the packet? Mizu should be unintrusive and should not affect the way packets flow.

Reply:
There is https://github.com/up9inc/mizu/blob/feat/afpacket/tap/source/handle_af_xdp.go#L141 which should send the packets back to the XDP socket. But upon taking another look, it might be not working as intended. What do you suggest?

Reply:
It kinda goes against the main appeal of using Mizu in the first place, as opposed to other options. Except for the cpu and mem consumption, it should not affect the user's system.
By rerouting the packets we add delays and introduce the risk of dropping packets in case of an error.

I suggest copying the packet and transmitting the copy to the Go code, the same way we do it in tlsTapper.

Reply:
Despite I searched for days or even weeks, I couldn't find a way to pass the packets to the sniffer without the `XDP_REDIRECT` involvement.

There are receive (RX) and transmit (TX) ring buffers in Linux. These buffers are shared buffers between the device driver and network interface card (NIC). According to my understanding we have to redirect the packets to an XDP socket, read the packets from that XDP socket and transmit them to somewhere such that they will continue their journey in the Linux network stack. The packets are actually captured inside [this method](https://github.com/asavie/xdp/blob/master/xdp.go#L423-L443).

Since the rebroadcast of the captured packets (transmit) is not working as intended, we might want to disable or remove `AF_XDP` from this PR.

Please also see;
- https://github.com/cloudflare/xdpcap/issues/49#issuecomment-813967235
- https://github.com/cloudflare/xdpcap/issues/32#issuecomment-563186616
- https://github.com/cloudflare/xdpcap/issues/51#issue-851220670
- https://github.com/cloudflare/xdpcap/issues/10#issuecomment-494408986

Reply:
1. In tls tapper we propagate packets (actually just the tls payload) from eBPF to Go with perf buffers. Why can't we do the same here?
2. What is the benefit of XDP? Why not remove it from this PR?